pkg/router/apisix.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	httpBackend := apisixRouteClone.Spec.HTTP[0]
+	httpBackend.Priority = maxPriority
+
+	_, primaryName, canaryName := canary.GetServiceNames()
+
+	primaryBackend := httpBackend.Backends[0]
+	primaryBackend.ServiceName = primaryName
+	primaryWeight := 100
+	primaryBackend.Weight = &primaryWeight
+	httpBackend.Backends[0] = primaryBackend
+
+	canaryWeight := 0
+	canaryBackend := a6v2.ApisixRouteHTTPBackend{
+		ServiceName:        canaryName,
+		ServicePort:        primaryBackend.ServicePort,
+		ResolveGranularity: primaryBackend.ResolveGranularity,
+		Weight:             &canaryWeight,
+		Subset:             primaryBackend.Subset,
+	}
+	httpBackend.Backends = append(httpBackend.Backends, canaryBackend)
+
+	apisixRouteClone.Spec.HTTP[0] = httpBackend
+	canaryApisixRouteName := fmt.Sprintf("%s-canary", canary.Spec.RouteRef.Name)
+	canaryApisixRoute, err := ar.apisixClient.ApisixV2().ApisixRoutes(canary.Namespace).Get(context.TODO(), canaryApisixRouteName, metav1.GetOptions{})
+
+	if errors.IsNotFound(err) {
+		route := &a6v2.ApisixRoute{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        canaryApisixRouteName,
+				Namespace:   canary.Namespace,
+				Annotations: apisixRouteClone.Annotations,
+				Labels:      apisixRouteClone.Labels,
+			},
+			Spec: apisixRouteClone.Spec,
+		}
+
+		if ar.setOwnerRefs {
+			route.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(canary, schema.GroupVersionKind{
+					Group:   flaggerv1.SchemeGroupVersion.Group,
+					Version: flaggerv1.SchemeGroupVersion.Version,
+					Kind:    flaggerv1.CanaryKind,
+				}),
+			}
+		}
+
+		_, err := ar.apisixClient.ApisixV2().ApisixRoutes(canary.Namespace).Create(context.TODO(), route, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("apisix route %s.%s create error: %w", route.Name, route.Namespace, err)
+		}
+
+		ar.logger.With("canary", fmt.Sprintf("%s.%s", canary.Name, canary.Namespace)).
+			Infof("apisix route %s.%s created", route.GetName(), canary.Namespace)
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("apisix route %s.%s query error: %w", canaryApisixRouteName, canary.Namespace, err)
+	}
+
+	diffHttpSpec := "placeholder"
+	diffBackend := "placeholder"
+	diffCanaryBackend := "placeholder"
+	if len(canaryApisixRoute.Spec.HTTP) == 1 &&
+		len(canaryApisixRoute.Spec.HTTP[0].Backends) == 2 {
+		diffHttpSpec = cmp.Diff(apisixRouteClone.Spec.HTTP[0],
+			canaryApisixRoute.Spec.HTTP[0],
+			cmpopts.IgnoreFields(a6v2.ApisixRouteHTTP{}, "Backends"))
+
+		diffBackend = cmp.Diff(apisixRouteClone.Spec.HTTP[0].Backends[0],
+			canaryApisixRoute.Spec.HTTP[0].Backends[0],
+			cmpopts.IgnoreFields(a6v2.ApisixRouteHTTPBackend{}, "Weight"))
+
+		diffCanaryBackend = cmp.Diff(apisixRouteClone.Spec.HTTP[0].Backends[1],
+			canaryApisixRoute.Spec.HTTP[0].Backends[1],
+			cmpopts.IgnoreFields(a6v2.ApisixRouteHTTPBackend{}, "Weight"))
+	}
+
+	if diffHttpSpec != "" || diffBackend != "" || diffCanaryBackend != "" {
```

Comment:
i don't understand the reason behind this. why can't we do something like:
```go
if diff := cmp.Diff(canaryApisixRoute.Spec, apixisRouteClone.Spec,
cmpopts.IgnoreFields(a6v2.ApisixRouteHTTPBackend{}, "Weight")
); diff != ""  {
```

Reply:
Great. Thanks for your advice.
When I was writing this part of the code I was confused and very unhappy with it, mainly because I wasn't familiar with the `cmp.Diff` API.

I will revise this part later.